webui/react/src/pages/ExperimentDetails/TrialCheckpoints.tsx

```diff
@@ -0,0 +1,418 @@
+        }}
+        id={record.uuid}
+        kind="checkpoint"
+        onError={handleError}
+        onTrigger={dropDownOnTrigger(record.uuid)}
+      />
+    );
+
+    const checkpointRenderer = (_: string, record: CoreApiGenericCheckpoint): React.ReactNode => {
+      return (
+        <Button
+          aria-label="View Checkpoint"
+          icon={<Icon name="checkpoint" showTooltip title="View Checkpoint" />}
+          onClick={() => handleOpenCheckpoint(record)}
+        />
+      );
+    };
+
+    const newColumns = [...defaultColumns].map((column) => {
+      column.sortOrder = null;
+      if (column.key === 'checkpoint') {
+        column.render = checkpointRenderer;
+      } else if (column.key === Checkpointv1SortBy.STATE) {
+        column.filterDropdown = stateFilterDropdown;
+        column.isFiltered = (settings) => !!(settings as Settings).state;
+        column.filters = Object.values(CheckpointState)
+          .filter((value) => value !== CheckpointState.Unspecified)
+          .map((value) => ({
+            text: <Badge state={value} type={BadgeType.State} />,
+            value,
+          }));
+      } else if (column.key === Checkpointv1SortBy.SEARCHERMETRIC) {
+        column.render = HumanReadableNumberRenderer;
+        column.title = `Searcher Metric (${experiment.config.searcher.metric})`;
+      } else if (column.key === 'actions') {
+        column.render = actionRenderer;
+      }
+      if (column.key === settings.sortKey) {
+        column.sortOrder = settings.sortDesc ? 'descend' : 'ascend';
+      }
+      return column;
+    });
+
+    return newColumns;
+  }, [
+    dropDownOnTrigger,
+    experiment.config.searcher.metric,
+    handleOpenCheckpoint,
+    settings.sortDesc,
+    settings.sortKey,
+    stateFilterDropdown,
+  ]);
+
+  const fetchTrialCheckpoints = useCallback(async () => {
+    if (!settings) return;
+
+    const states = settings.state?.map((state) => encodeCheckpointState(state as CheckpointState));
+    try {
+      const response = await getTrialCheckpoints(
+        {
+          id: trial.id,
+          limit: settings.tableLimit,
+          offset: settings.tableOffset,
+          orderBy: settings.sortDesc ? 'ORDER_BY_DESC' : 'ORDER_BY_ASC',
+          sortBy: validateDetApiEnum(Checkpointv1SortBy, settings.sortKey),
+          states: validateDetApiEnumList(Checkpointv1State, states),
+        },
+        { signal: canceler.signal },
+      );
+      setTotal(response.pagination.total ?? 0);
+      if (!isEqual(response.checkpoints, checkpoints)) {
+        setCheckpoints(response.checkpoints);
+      }
+    } catch (e) {
+      handleError(e, {
+        publicSubject: `Unable to fetch ${f_flat_runs ? 'run' : 'trial'} ${trial.id} checkpoints.`,
+        silent: true,
+        type: ErrorType.Api,
+      });
+    } finally {
+      setIsLoading(false);
+    }
+  }, [f_flat_runs, settings, trial.id, canceler.signal, checkpoints]);
```

Comment:
this has a circular reference where we update the callback reference when `checkpoints` changes but we set `checkpoints` in the function. because `usePolling` is set up to restart the poll when the polling function changes, it'll cause the component to send multiple requests. use the functional updater here:
```suggestion
      setCheckpoints((cps) => {
        return isEqual(response.checkpoints, cps) ? response.checkpoints : cps;
      })
    } catch (e) {
      handleError(e, {
        publicSubject: `Unable to fetch ${f_flat_runs ? 'run' : 'trial'} ${trial.id} checkpoints.`,
        silent: true,
        type: ErrorType.Api,
      });
    } finally {
      setIsLoading(false);
    }
  }, [f_flat_runs, settings, trial.id, canceler.signal]);
```